profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp

```diff
@@ -165,4 +271,224 @@ bool ClrEventsParser::TryGetEventInfo(LPCBYTE pMetadata, ULONG cbMetadata, WCHAR
+    uint64_t pauseDuration,
+    uint64_t totalDuration,
+    uint64_t endTimestamp
+    )
+{
+    if (_pGarbageCollectionsListener != nullptr)
+    {
+        _pGarbageCollectionsListener->OnGarbageCollection(
+            number,
+            generation,
+            reason,
+            type,
+            isCompacting,
+            pauseDuration,
+            totalDuration,
+            endTimestamp
+            );
+    }
+}
+
+GCDetails& ClrEventsParser::GetCurrentGC()
+{
+    if (_gcInProgress.Number != -1)
+    {
+        return _gcInProgress;
+    }
+
+    return _currentBGC;
+}
+
+void ClrEventsParser::ClearCollections()
+{
+    ResetGC(_currentBGC);
+    ResetGC(_gcInProgress);
+}
+
+void ClrEventsParser::ResetGC(GCDetails& gc)
+{
+    gc.Number = -1;
+    gc.Generation = 0;
+    gc.Reason = (GCReason)0;
+    gc.Type = (GCType)0;
+    gc.IsCompacting = false;
+    gc.PauseDuration;
+    gc.StartTimestamp = 0;
+}
+
+void ClrEventsParser::InitializeGC(GCDetails& gc, GCStartPayload& payload)
+{
+    gc.Number = payload.Count;
+    gc.Generation = payload.Depth;
+    gc.Reason = (GCReason)payload.Reason;
+    gc.Type = (GCType)payload.Type;
+    gc.IsCompacting = false;
+    gc.PauseDuration = 0;
+    gc.StartTimestamp = GetCurrentTimestamp();
+}
+
+void ClrEventsParser::OnGCTriggered()
+{
+    // all previous collections are finished
+    ClearCollections();
+}
+
+void ClrEventsParser::OnGCStart(GCStartPayload& payload)
+{
+    // If a BCG is already started, FGC (0/1) are possible and will finish before the BGC
+    //
+    if ((payload.Depth == 2) && (payload.Type == GCType::BackgroundGC))
+    {
+        InitializeGC(_currentBGC, payload);
+    }
+    else
+    {
+        InitializeGC(_gcInProgress, payload);
+    }
+}
+
+void ClrEventsParser::OnGCStop(GCEndPayload& payload)
+{
+}
+
+void ClrEventsParser::OnGCSuspendEEBegin()
+{
+    // we don't know yet what will be the next GC corresponding to this suspension
+    // so it is kept until next GCStart
+    _suspensionStart = GetCurrentTimestamp();
+}
+
+void ClrEventsParser::OnGCRestartEEEnd()
+{
+    GCDetails& gc = GetCurrentGC();
+    if (gc.Number == -1)
+    {
+        // this might happen (seen in workstation + concurrent mode)
+        // --> just skip the suspension because we can associate to a GC
+        _suspensionStart = 0;
+        return;
+    }
+
+    // compute suspension time
+    uint64_t suspensionDuration = 0;
+    uint64_t currentTimestamp = GetCurrentTimestamp();
+    if (_suspensionStart != 0)
+    {
+        suspensionDuration = currentTimestamp - _suspensionStart;
+        NotifySuspension(gc.Number, gc.Generation, suspensionDuration, currentTimestamp);
+
+        _suspensionStart = 0;
+    }
+    else
+    {
+        // bad luck: a xxxBegin event has been missed
+        LOG_GC_EVENT("### missing Begin event");
+    }
+    gc.PauseDuration += suspensionDuration;
+
+    // could be the end of a gen0/gen1 or of a non concurrent gen2 GC
+    if (
```

Comment:
why not having the expression on the same line ?

Reply:
✅